comp/core/secrets/secretsimpl/secrets.go

```diff
@@ -53,14 +80,72 @@ type secretContext struct {
 	yamlPath string
 }
 
-func init() {
-	secretCache = make(map[string]string)
-	secretOrigin = make(handleToContext)
+// TODO: Hack to maintain a singleton reference to the secrets Component
+//
+// Only needed temporarily, since the secrets.Component is needed for the diagnose functionality.
+// It is very difficult right now to modify diagnose because it would require modifying many
+// function signatures, which would only increase future maintenance. Once diagnose is better
+// integrated with Components, we should be able to remove this hack.
+//
+// Other components should not copy this pattern, it is only meant to be used temporarily.
+var mu sync.Mutex
+var instance *secretResolver
+
+func newEnabledSecretResolver() *secretResolver {
+	return &secretResolver{
+		cache:   make(map[string]string),
+		origin:  make(handleToContext),
+		enabled: true,
+	}
 }
 
-func registerSecretOrigin(handle string, origin string, yamlPath []string) {
+func newSecretResolverProvider(deps dependencies) provides {
+	resolver := newEnabledSecretResolver()
+	resolver.enabled = deps.Params.Enabled
+
+	{
+		mu.Lock()
+		defer mu.Unlock()
+		if instance == nil {
+			instance = resolver
+		}
+	}
+
+	return provides{
+		Comp:          resolver,
+		FlareProvider: flaretypes.NewProvider(resolver.fillFlare),
+	}
+}
+
+func (r *secretResolver) IsEnabled() bool {
+	return r.enabled
+}
+
+// GetInstance returns the singleton instance of the secret.Component
+func GetInstance() secrets.Component {
+	mu.Lock()
+	defer mu.Unlock()
+	if instance == nil {
+		deps := dependencies{Params: secrets.Params{Enabled: true}}
+		p := newSecretResolverProvider(deps)
+		instance = p.Comp.(*secretResolver)
+	}
+	return instance
+}
+
```

Comment:
I feel that this should return a disabled secret object. Unless the main directly enabled it we should fallback to the default which is disabled.

Also, in which case would the instance be nil ? Is is only for serverless ? I so we might want a dedicated method for it that let it configure the comp without FX. WDYT ?

Reply:
Disabled by default makes more sense, good call.

`secretsimpl.GetInstance()` is only needed for `diagnoseChecksInCLIProcess`. The `instance` var will only be nil here if the main entry point didn't Provide the secret Component, but the current code makes that impossible. Perhaps this should be changed to `panic` instead?

Reply:
:thinking: panics are always a bit scary in non vital component. I feel like returning a disabled secret would be the best option, WDYT ?